alf/environments/suite_carla.py

```diff
@@ -0,0 +1,1163 @@
+            self._server = CarlaServer(rpc_port, streaming_port)
+
+        self._batch_size = batch_size
+        self._world = None
+        try:
+            for i in range(10):
+                try:
+                    logging.info(
+                        "Waiting for server to start. Try %d" % (i + 1))
+                    self._client = carla.Client("localhost", rpc_port)
+                    self._world = self._client.load_world(map_name)
+                    break
+                except RuntimeError:
+                    continue
+        finally:
+            if self._world is None:
+                self._server.stop()
+                assert self._world is not None, "Fail to start server."
+
+        logging.info("Server started.")
+        self._alf_world = World(self._world)
+        self._safe = safe
+        self._vehicle_filter = vehicle_filter
+
+        settings = self._world.get_settings()
+        settings.synchronous_mode = True
+        settings.fixed_delta_seconds = step_time
+
+        self._world.apply_settings(settings)
+        self._map_name = map_name
+        self._spawn()
+
+    def _spawn(self):
+        blueprints = self._world.get_blueprint_library().filter(
+            self._vehicle_filter)
+        assert len(
+            blueprints) > 0, "Cannot find vehicle '%s'" % self._vehicle_filter
+        if self._safe:
+            blueprints = [
+                x for x in blueprints
+                if int(x.get_attribute('number_of_wheels')) == 4
+            ]
+            blueprints = [x for x in blueprints if not x.id.endswith('isetta')]
+            blueprints = [
+                x for x in blueprints if not x.id.endswith('carlacola')
+            ]
+            blueprints = [
+                x for x in blueprints if not x.id.endswith('cybertruck')
+            ]
+            blueprints = [x for x in blueprints if not x.id.endswith('t2')]
+        assert len(
+            blueprints
+        ) > 0, "Cannot find safe vehicle '%s'" % self._vehicle_filter
+
+        spawn_points = self._world.get_map().get_spawn_points()
+        number_of_spawn_points = len(spawn_points)
+        if self._batch_size < number_of_spawn_points:
+            random.shuffle(spawn_points)
+        elif self._batch_size > number_of_spawn_points:
+            raise ValueError(
+                "requested %d vehicles, but could only find %d spawn points" %
+                (self._batch_size, number_of_spawn_points))
+
+        SpawnActor = carla.command.SpawnActor
+
+        commands = []
+        for transform in spawn_points[:self._batch_size]:
+            blueprint = random.choice(blueprints)
+            if blueprint.has_attribute('color'):
+                color = random.choice(
+                    blueprint.get_attribute('color').recommended_values)
+                blueprint.set_attribute('color', color)
+            if blueprint.has_attribute('driver_id'):
```

Comment:
What's the purpose of having random driver ids?

Reply:
Actually, I don't know what effect driver id has. Cannot find documentation about it. Carla source code doesn't say it either.